File: prepare_radboud_data_for_training.py
"""Python version of matlab script of same name.

Script does image formation on ultrasound acquisition data.

Author: Ryan A.L. Schoop
"""

from os import makedirs
from os.path import join
from glob import glob
from time import time
from datetime import timedelta
import numpy as np
from scipy.io import loadmat
from h5py import File
from torch import from_numpy, real, zeros
import hdf5storage
from deepus import torch_fkmig, torch_image_formation

# Script assumes that the data is stored in 'ExperimentalData/CIRS073_RUMC' or
# 'ExperimentalData/CIRS040GSE/' within some storage_path root directory.
storage_path = '/export/scratch2/felix/Dropbox/Data/US/DeepUSData'
data_set = 'CIRS073_RUMC' # Pick CIRS073_RUMC or CIRS040GSE
experimental_data_path = join(storage_path, 'ExperimentalData', data_set)

# This is where the training data will be stored.
training_data_path_root = join(storage_path, 'TrainingData', data_set)

# Settings for reconstruction
# Sub-sampled number of steering angles used for reconstruction [1, ..., 75]
n_ang_ss = 1
# Truncation of pixels in axial direction.
# Note: this is rather replaced by proper interpolation.
nz_cutoff = 1500

# Load the header containing information about the transducer settings.
if data_set == 'CIRS040GSE':
    header_file = join(experimental_data_path, 'USHEADER_20220330105548.mat')
elif data_set == 'CIRS073_RUMC':
    header_file = join(experimental_data_path, 'USHEADER_20210624112712S.mat')
else:
    raise ValueError('Unsupported data set.')
usheader = loadmat(header_file, struct_as_record=False)['USHEADER'][0][0]

n_ang = len(usheader.xmitAngles)
n_ele = usheader.xmitDelay.shape[1]

# Prepare to loop over all data sets
data_files = glob(join(experimental_data_path, '**', 'USDATA_*.mat'),
                  recursive=True)
n_data = len(data_files)
# Result directory for images.
training_data_path_img = join(training_data_path_root, 'Images',
                              f'nAng{n_ang_ss}')
makedirs(training_data_path_img, exist_ok=True)
# Result directory for training.
training_data_path_data = join(training_data_path_root, 'Data',
                               f'nAng{n_ang_ss}')
makedirs(training_data_path_data, exist_ok=True)

# Load single dataset for some overarching information.
with File(data_files[1]) as f:
    # Disregard empty dimensions, get right shape and convert to double.
    usdata = np.swapaxes(np.squeeze(np.array(f.get('USDATA'))),
                         0, 2).astype(np.float64)

# If specified, subsample the amount of angles.
if n_ang_ss == 1:
    ang_ind = np.squeeze(np.argwhere(usheader.xmitAngles == 0))[0]
else:
    ang_ss = np.linspace(np.squeeze(usheader.xmitAngles[0]),
                         np.squeeze(usheader.xmitAngles[-1]), n_ang_ss)
    ang_ind = np.zeros(n_ang_ss, dtype=np.int64)
    for i_ang in range(n_ang_ss):
        ang_ind[i_ang] = np.argmin(np.abs(np.squeeze(usheader.xmitAngles)
                                          - ang_ss[i_ang])).squeeze()

usdata = np.atleast_3d(usdata[:, :, ang_ind])
usheader.xmitAngles = np.atleast_1d(np.squeeze(usheader.xmitAngles)[ang_ind])
usheader.xmitDelay = np.squeeze(usheader.xmitDelay)[ang_ind, :]
usheader.xmitFocus = np.squeeze(usheader.xmitFocus)[ang_ind]
usheader.xmitApodFunction = np.squeeze(usheader.xmitApodFunction)[ang_ind, :]

# Speed of sound used for calculations.
sos_bgn = np.squeeze(usheader.c)
# Used to correct for general delay of the acoustic lens.
lens_delay = 96

# Main loop
for i, data_file in enumerate(data_files):
    print(f'Reconstructing data file index {i}...')
    rc_time_start = time()

    with File(data_file) as f:
        usdata = np.swapaxes(np.squeeze(np.array(f.get('USDATA'))),
                            0, 2).astype(np.float64)
        usdata = np.atleast_3d(usdata[:, :, ang_ind])

    # Correct general delay, i.e. lens correction
    usdata = usdata[lens_delay:, :, :]

    # Correct angle dependent delay
    n_t = usdata.shape[0]
    dt = 1 / np.squeeze(usheader.fs)
    delay = np.abs((n_ele - 1) / 2 * np.squeeze(usheader.pitch)
                   * np.sin(np.deg2rad(usheader.xmitAngles)) / sos_bgn)
    delay_ind = np.floor(delay / dt).astype(int)
    for i_ang in range(n_ang_ss):
        usdata[:(n_t - delay_ind[i_ang]), :, i_ang] = (
            usdata[delay_ind[i_ang]:, :, i_ang])
        usdata[(n_t - delay_ind[i_ang]):, :, i_ang] = 0

    # Correct channel 125
    usdata[:, 125, :] = 2 * usdata[:, 125, :]

    img_fkmig = zeros((usdata.shape[0], usdata.shape[1]))
    for k, xmitAngle in enumerate(usheader.xmitAngles):
        # Set up fk-migration and run it.
        fk_para = {
            'pitch': np.squeeze(usheader.pitch).astype(np.float32),
            'fs': np.squeeze(usheader.fs).astype(np.float32),
            'tx_angle': np.deg2rad(xmitAngle).astype(np.float32),
            'c': sos_bgn.astype(np.float32),
            't0': 0
        }
        # Compounding
        img_fkmig = ((k * img_fkmig
                      + torch_fkmig(from_numpy(usdata[:, :, k]), fk_para))
                     / (k + 1))

    # Image post processing
    img_pp = torch_image_formation(real(img_fkmig), -70)

    # Convert back to numpy.
    img_fkmig = img_fkmig.detach().cpu().numpy()
    img_pp = img_pp.detach().cpu().numpy()

    # Truncation in axial direction, though rather done by interpolation.
    img_fkmig = img_fkmig[:nz_cutoff, :]
    img_pp = img_pp[:nz_cutoff, :]

    # Saving.
    options = hdf5storage.Options(matlab_compatible=True)
    res_filename_img = join(training_data_path_img, f'img_{i + 1}.mat')
    hdf5storage.writes(
        {'/img_fkmig': img_fkmig, '/img_pp': img_pp, '/fk_para': fk_para},
        res_filename_img, options=options)

    data = usdata
    res_filename_data = join(training_data_path_data, f'data_{i + 1}.mat')
    fk_para['tx_angle'] = usheader.xmitAngles
    hdf5storage.writes({'/data': data, '/fk_para': fk_para}, res_filename_data,
                       options=options)

    rc_time_end = time()
    print('Reconstruction Duration: '
          f'{str(timedelta(seconds=rc_time_end - rc_time_start))}')

# Additional metadata
md_usheader = loadmat(header_file, matlab_compatible=True,
                      variable_names=['USHEADER'])['USHEADER']

if data_set == 'CIRS040GSE':
    lesion_idx = {
        'high_attenuation_hypoechoic': np.arange(1, 6).astype(np.float64),
        'high_attenuation_wires': np.arange(6, 11).astype(np.float64),
        'high_attenuation_plusdb': np.arange(11, 16).astype(np.float64),
        'high_attenuation_minusdb': np.arange(16, 21).astype(np.float64),
        'low_attenuation_hypoechoic': 20 + np.arange(1, 6).astype(np.float64),
        'low_attenuation_wires': 20 + np.arange(6, 11).astype(np.float64),
        'low_attenuation_plusdb': 20 + np.arange(11, 16).astype(np.float64),
        'low_attenuation_minusdb': 20 + np.arange(16, 21).astype(np.float64),
        'nIdx': float(n_data)
    }
elif data_set == 'CIRS073_RUMC':
    keys = ('hyper_lesion1', 'hyper_lesion2', 'hyper_lesion3',
                       'hypo_lesion1', 'hypo_lesion2', 'hypo_lesion3',
                       'no_lesion1', 'no_lesion2')
    file_ids = (join('hyper_lesion', 'USDATA'), 'hyper_lesion2',
                   'hyper_lesion3', join('hypo_lesion', 'USDATA'),
                   'hypo_lesion2', 'hypo_lesion3',
                   join('no_lesion', 'USDATA'), 'no_lesion2')
    lesion_idx = {
        key: np.nonzero([file_id in data_file
                         for data_file in data_files
                         ])[0].astype(np.float64) + 1
             for key, file_id in zip(keys, file_ids)
    }
    lesion_idx['nIdx'] = float(n_data)
else:
    raise ValueError("Unsupported data set code.")

if n_ang_ss == 1:
    # The deepus.py library currently only supports loading headers of
    # one subsampled angle. If the metadata file has a USHEADER
    # configuration of more than one angle then the deepus.py
    # load_header function will fail. This condition will ensure only
    # the case of one subsampled angle is saved.

    md_usheader['xmitAngles'] = usheader.xmitAngles
    # Indexing to get to the actual value stored in the data object.
    md_usheader['xmitDelay'][0][0] = usheader.xmitDelay
    md_usheader['xmitFocus'] = usheader.xmitFocus
    md_usheader['xmitApodFunction'][0][0] = usheader.xmitApodFunction.astype(
        np.float64)

    save_dict = {
        '/USHEADER': md_usheader,
        '/TargetInfo': {'nz_cutoff': nz_cutoff},
        '/LesionIdx': lesion_idx
    }

    res_filename_metadata = join(training_data_path_root, 'metadata.mat')
    hdf5storage.writes(save_dict, res_filename_metadata,
                    options=hdf5storage.Options(matlab_compatible=True))

File: matlab/scripts/prepare_radboud_data_for_training.m
% script to study the first experimental data aquired by Gijs Hendriks
% <Gijs.Hendriks@radboudumc.nl> on 24 June 2021
%
% author: Felix Lucka, Georgios Pilikos, Ryan Schoop
% date:        01.09.2021
% last update: 20.06.2023


clear
clc
%close all

% initializes the toolbox
startup

% the script assumes that the data is stored in:
% storage_path/ExperimentalData/CIRS073_RUMC/, or
% storage_path/ExperimentalData/CIRS040GSE/

data_set_name = 'CIRS073_RUMC';
experimental_data_path = [storage_path 'ExperimentalData' fs data_set_name];

% this is where the prepared training data will be stored
training_data_path_root = [storage_path 'TrainingData' fs data_set_name];

%% settings of the reconstruction

n_ang_ss      = 1; % sub-sampled number of steering angles used in the reconstruction (1,...,75)
nz_cutoff     = 1500; % cut off in pixels depth direction (TODO: replace by proper interpolation)
%passband_fac  = 1; % relative filtering bandwidth

%% load the header

% the script tries to load the data header which contains all the
% information about the transducer settings

% Pick the header file based on the data_set chosen above.
header_file = [experimental_data_path fs 'USHEADER_20210624112712.mat'];
load(header_file);
USHEADER

n_ang = length(USHEADER.xmitAngles);
n_ele = size(USHEADER.xmitDelay, 2); % number of transducer elements

%% prepare the loop over all data sets

% get a list of all data files
data_files = rDir(experimental_data_path, 'USDATA_*.mat');
n_data     = length(data_files)

% make res dir for image
% training_data_path_img = [training_data_path_root fs 'Images' fs 'PB' num2str(passband_fac) fs 'nAng' int2str(n_ang_ss) fs];
training_data_path_img = [training_data_path_root fs 'Images' fs 'nAng' int2str(n_ang_ss) fs];
makeDir(training_data_path_img);
% make res dir for data
training_data_path_data = [training_data_path_root fs 'Data' fs 'nAng' int2str(n_ang_ss) fs];
makeDir(training_data_path_data);

% load first data set
load([data_files(2).folder fs data_files(2).name]);
    
% convert data to double and get rid of empty dimensions
USDATA = double(squeeze(USDATA));
    
% sub-sample in angle if we want to
if(n_ang_ss == 1)
    ang_ind = find(USHEADER.xmitAngles == 0);
else
    ang_ss  = linspace(USHEADER.xmitAngles(1), USHEADER.xmitAngles(end), n_ang_ss);
    ang_ind = zeros(n_ang_ss, 1);
    for i_ang = 1:n_ang_ss
        [~,ang_ind(i_ang)] = min(abs(USHEADER.xmitAngles - ang_ss(i_ang)));
    end
end
    
USDATA                    = USDATA(:,:,ang_ind);
USHEADER.xmitAngles       = USHEADER.xmitAngles(ang_ind);
USHEADER.xmitDelay        = USHEADER.xmitDelay(ang_ind,:);
USHEADER.xmitFocus        = USHEADER.xmitFocus(ang_ind);
USHEADER.xmitApodFunction = USHEADER.xmitApodFunction(ang_ind,:);
        
% sound speed used for computations
sos_bgn     = USHEADER.c;
lens_delay  = 96;

%% main loop
for i_data = 1:n_data
    
    fprintf(['reconstructing data set ' int2str(i_data) '...'])
    clock_cmp = tic;
    
    load([data_files(i_data).folder fs data_files(i_data).name]);
    
    % convert data to double and get rid of empty dimensions
    USDATA = double(squeeze(USDATA));
    USDATA = USDATA(:,:,ang_ind);

    
    % data pre-processing
    
    % correct general delay aka "lens correction"
    USDATA      = USDATA(lens_delay:end, :, :);
    n_t         = size(USDATA, 1);
    
    % correct angle-dependent delay
    dt        = 1/USHEADER.fs;
    delay_1   = USHEADER.xmitDelay(:,1) - (mean(USHEADER.xmitDelay(:,64:65),2));
    delay_2   = abs((n_ele-1)/2*USHEADER.pitch * sin(deg2rad(USHEADER.xmitAngles)) / sos_bgn);
    % choose a way to compute the delay
    delay     = delay_2;
    delay_ind = floor(delay./dt) + 1;
    for i_ang=1:n_ang_ss
        USDATA(1:n_t-delay_ind(i_ang)+1, :,i_ang)   = USDATA(delay_ind(i_ang):end, :,i_ang);
        USDATA(n_t-delay_ind(i_ang)+2:end, :,i_ang) = 0;
    end
    
    % correct channel 125
    USDATA(:,125,:) = 2*USDATA(:,125,:);
    
    % set up F-K migration and run it
    fk_para          = [];
    fk_para.TXangle  = deg2rad(USHEADER.xmitAngles(:));
    fk_para.pitch    = USHEADER.pitch;
    fk_para.t0       = 0;
    fk_para.c        = sos_bgn;
    fk_para.fs       = USHEADER.fs;
    %passband         = (USHEADER.fc + (USHEADER.fc * USHEADER.fbw * passband_fac*[-0.5,0.5])) / (USHEADER.fs/2);
    %fk_para.passband = passband;
    
    % call F-K migration
    [img_fkmig, fk_para] = fkmig(USDATA, fk_para);
    
    % image post processing
    img_pp = img_fkmig;
    % hilbert transform
    img_pp = abs(hilbert(img_pp));
    % % SNR transform
    img_pp = 20*log10(img_pp);
    img_pp = img_pp - max(img_pp(:));
    % clipping
    img_pp(img_pp < -70) = -70;
    img_pp(img_pp > 0) = 0;
    % shift from negative values
    img_pp = (img_pp + 70) / 70;
    
    % TODO: reintroduce the interpolation
    % interpolate images to isotropic grid twice as deep as wide
    %x_vec = fk_para.x(:)';
    %dx    = x_vec(2) - x_vec(1);
    %z_vec = dx * (0:(2*length(x_vec)-1));
    %img_fkmig = interpn(fk_para.z, fk_para.x, img_fkmig, z_vec(:), x_vec);
    %img_pp    = interpn(fk_para.z, fk_para.x, img_pp, z_vec(:), x_vec);
    img_fkmig = img_fkmig(1:nz_cutoff,:);
    img_pp    = img_pp(1:nz_cutoff,:);
    x_vec = fk_para.x(:)';
    z_vec = fk_para.z(1:nz_cutoff)';
    
    % save images
    res_filename_img = [training_data_path_img 'img_' int2str(i_data) '.mat'];
    save(res_filename_img, 'img_fkmig', 'img_pp', 'fk_para', 'x_vec', 'z_vec','-v7.3');
    
    % save data
    data = USDATA;
    res_filename_data = [training_data_path_data 'data_' int2str(i_data) '.mat'];
    save(res_filename_data, 'data', 'fk_para','-v7.3');
    
    disp(['done. Computation time:' convertSec(toc(clock_cmp))])
    
end

%% Extra metadata
if strcmp(data_set_name,'CIRS073_RUMC')
    LesionIdx.hyper_lesion1 = find(endsWith({data_files.folder},'hyper_lesion'));
    LesionIdx.hyper_lesion2 = find(contains({data_files.folder},'hyper_lesion2'));
    LesionIdx.hyper_lesion3 = find(contains({data_files.folder},'hyper_lesion3'));
    LesionIdx.hypo_lesion1 = find(endsWith({data_files.folder},'hypo_lesion'));
    LesionIdx.hypo_lesion2 = find(contains({data_files.folder},'hypo_lesion2'));
    LesionIdx.hypo_lesion3 = find(contains({data_files.folder},'hypo_lesion3'));
    LesionIdx.no_lesion1 = find(endsWith({data_files.folder},'no_lesion'));
    LesionIdx.no_lesion2 = find(contains({data_files.folder},'no_lesion2'));
elseif strcmp(data_set_name,'CIRS040GSE')
    LesionIdx.high_attenuation_hypoechoic = (1:5);
    LesionIdx.high_attenuation_wires = (6:10);
    LesionIdx.high_attenuation_plusdb = (11:15);
    LesionIdx.high_attenuation_minusdb = (16:20);
    LesionIdx.low_attenuation_hypoechoic = 20 + (1:5);
    LesionIdx.low_attenuation_wires = 20 + (6:10);
    LesionIdx.low_attenuation_plusdb = 20 + (11:15);
    LesionIdx.low_attenuation_minusdb = 20 + (16:20);
else
    error('Unknown data_set specified.')
end
LesionIdx.nIdx = n_data; % Don't forget this one!!

TargetInfo.nz_cutoff = nz_cutoff;

res_filename_metadata = [training_data_path_root fs 'metadata' '.mat'];

save(res_filename_metadata, 'USHEADER', 'TargetInfo', 'LesionIdx', '-v7.3');